File: JavaScript/Sorting Algorithms/bubbleSort.js
function swap(arr, idx1, idx2) {
    let temp = arr[idx1];
    arr[idx1] = arr[idx2];
    arr[idx2] = temp;
    return arr;
}

function bubbleSort(arr) {
    let sortedArray = arr;
    let loops = 0;
    for (var i = sortedArray.length - 1; i >= 0; i--) {
        for (var j = 0; j < i; j++){
            loops++;
            if (sortedArray[j] > sortedArray[j+1]){
                swap(sortedArray, j, j+1);
            }
        }
    }
    console.log(loops);
    return sortedArray;
}

function bubbleSortOpt(arr) {
    let sortedArray = arr;
    let noSwaps;
    let loops = 0;
    for (var i = sortedArray.length - 1; i >= 0; i--) {
        noSwaps = true;
        for (var j = 0; j < i; j++){
            loops++;
            if (sortedArray[j] > sortedArray[j+1]){
                swap(sortedArray, j, j+1);
                noSwaps = false;
            }
        }
        if (noSwaps) {
            console.log(loops);
            return sortedArray;
        }
    }
}

//console.log(swap([2,5,1,34,53,62],3,4))
console.log(bubbleSort([100,50,25,10,5, 16]))
console.log(bubbleSortOpt([100,50,25,10,5, 16]))

console.log(bubbleSort([2,3,4,5,6]))
console.log(bubbleSortOpt([2,3,4,5,6]))

File: Python/Sorting Algorithms/bubbleSort.py
def bubbleSort(arr):
    sortedArray = arr
    loops = 0
    for i in range(len(arr) -1, -1, -1):
        noSwaps = True
        for j in range(0,i):
            loops += 1
            if sortedArray[j] > sortedArray[j+1]:
                sortedArray[j], sortedArray[j+1] = sortedArray[j+1], sortedArray[j]
                noSwaps = False
        if noSwaps:
            break
    return sortedArray, loops

print(bubbleSort([5,4,3,2,1]))
print(bubbleSort([1,2,3,4,5]))

print(bubbleSort([100,50,25,10,5, 16]))

print(bubbleSort([2,3,4,5,6]))